guides/11_other-tutorials/create-immersive-demo.md

```diff
@@ -0,0 +1,199 @@
+                    async for response in turn:
+                        if data := response.data:
+                            audio = np.frombuffer(data, dtype=np.int16).reshape(1, -1)
+                        self.audio_queue.put_nowait(audio)
+                except websockets.exceptions.ConnectionClosedOK:
+                    print("connection closed")
+                    break
+
+    # Video: receive and (optionally) send frames to Gemini
+    async def video_receive(self, frame: np.ndarray):
+        self.video_queue.put_nowait(frame)
+        if self.session and (time.time() - self.last_frame_time > 1.0):
+            self.last_frame_time = time.time()
+            await self.session.send(input=encode_image(frame))
+            # If there is an uploaded image passed alongside the WebRTC component,
+            # it will be available in latest_args[2]
+            if self.latest_args[2] is not None:
+                await self.session.send(input=encode_image(self.latest_args[2]))
+
+    async def video_emit(self) -> np.ndarray:
+        frame = await wait_for_item(self.video_queue, 0.01)
+        if frame is not None:
+            return frame
+        # Fallback while waiting for first frame
+        return np.zeros((100, 100, 3), dtype=np.uint8)
+
+    # Audio: forward microphone audio to Gemini
+    async def receive(self, frame: tuple[int, np.ndarray]) -> None:
+        _, array = frame
+        array = array.squeeze()  # (num_samples,)
+        audio_message = encode_audio(array)
+        if self.session:
+            await self.session.send(input=audio_message)
+
+    # Audio: emit Gemini’s audio back to the client
+    async def emit(self):
+        array = await wait_for_item(self.audio_queue, 0.01)
+        if array is not None:
+            return (self.output_sample_rate, array)
+        return array
+
+    async def shutdown(self) -> None:
+        if self.session:
+            self.quit.set()
+            await self.session.close()
+            self.quit.clear()
+```
+
+
+## 3) Setup Stream and Gradio UI
+We’ll add an optional `gr.Image` input alongside the `WebRTC` component. The handler will access this in `self.latest_args[1]` when sending frames to Gemini.
+
+```python
+import gradio as gr
+from fastrtc import Stream, WebRTC, get_hf_turn_credentials
+
+
+stream = Stream(
+    handler=GeminiHandler(),
+    modality="audio-video",
+    mode="send-receive",
+    server_rtc_configuration=get_hf_turn_credentials(ttl=600*10000),
+    rtc_configuration=get_hf_turn_credentials(),
+    additional_inputs=[
+        gr.Markdown(
+            "## 🎨 Art Critic\n\n"
+            "Provide an image of your artwork and Gemini will critique it for you."
+            "To get a Gemini API key, please visit the [Gemini API Key](https://console.cloud.google.com/apis/api/generativelanguage.googleapis.com/credentials) page."
+            "To get an HF Token, please visit the [HF Token](https://huggingface.co/settings/tokens) page."
```

Comment:
What permissions does the HF token need? Would be great to specify so that someone can generate a fine-grained token.